library/src/compute/calc.rs

```diff
@@ -404,6 +408,170 @@ pub fn log(
     Value::Float(result)
 }
 
+/// Calculates the product of a range of numbers. Used to calculate permutations.
+/// Returns None if the result is larger than `u64::MAX`
+fn factorial_range(start: u64, end: u64) -> Option<u64> {
+    // By convention
+    if end + 1 < start {
+        return Some(0);
+    }
+
+    let mut count: u64 = 1;
+    let real_start: u64 = cmp::max(1, start);
+
+    for i in real_start..=end {
+        match count.checked_mul(i) {
+            None => return None,
+            Some(s) => count = s,
+        };
+    }
+    Some(count)
+}
+
+/// Calculates the factorial of an integer.
+/// Returns `None` if the result is larger than `u64::MAX`
+fn factorial(number: u64) -> Option<u64> {
+    factorial_range(1, number)
+}
+
+/// Calculate the factorial of a number.
+///
+/// ## Example
+/// ```example
+/// #calc.fact(5)
+/// ```
+///
+/// Display: Factorial
+/// Category: calculate
+/// Returns: integer
+#[func]
+pub fn fact(
+    /// The number whose factorial to calculate. Must be positive.
+    value: Spanned<Num>,
+) -> Value {
+    let number = match value.v {
+        Num::Float(_) => {
+            bail!(value.span, "a factorial argument must be an integer");
+        }
+        // Note: 20 is a hardcoded limit, as 21! > i64::MAX
+        Num::Int(i) if i > 20 => {
+            bail!(value.span, "a factorial argument should not exceed 20")
+        }
+        Num::Int(i) if i < 0 => {
+            bail!(value.span, "a factorial argument must be positive")
+        }
+        Num::Int(i) => i,
+    } as u64;
+
+    // Safe unwrap because we checked earlier that the factorial was less than the overflow limit, i64::MAX
+    Value::Int(factorial(number).unwrap_or_default() as i64)
+}
+
+/// Checks to make sure that a given `Num` is a positive integer.
+macro_rules! check_positive_integer_argument {
```

Comment:
Not necessary if you just use `NonZeroI64` everywhere instead of `Num` :)

Reply:
This is not yet addressed, right?

Reply:
You can replace `i64` with `NonZeroI64` wherever you want a positive integer.

Reply:
By positive I mean not strictly positive, it can be zero too

Reply:
Basically it should be an `u64` which is less than `i64::MAX` because it is then converted to an `i64`

Reply:
Ah I see, then `u64`. There isn't yet a `Cast` implementation for it, but you can easily add it in src/eval/cast.rs next to the others. Should be almost the same as the `cast_from_value!` and `cast_to_value!` for `usize`.